### PR TITLE
_pixelbuf: Increase performance of brightness-scaling

### PR DESCRIPTION
**Pull request by jepler**
On the Pico, this increases the "fill rate" of
```python
    pixels[:] = newvalues
```
considerably.  On a strip of 240 RGB LEDs, auto_write=False, the timingsare:

<table>
<tr> <td> Brightness <td> Before <td> After <td> Improvement
<tr> <th> 1.0        <td> 117 kpix/s <td> 307 kpix/s <td> 2.62x
<tr> <th> 0.07       <td> 117 kpix/s <td> 273 kpix/s <td> 2.33x
</table>

It's worth noting that even the "before" rate is fast compared to the time to transmit a single neopixel, but any time we can gain back in the whole pipeline will let marginal animations work a little better. To set all the pixels in this way and then show() gives a pleasant bump to the framerate, from about 108Hz to 124Hz (1.15x)

The main source of speed-up is using integer math instead of floating point math for the calculation of the post-scaled pixel values.  A slight secondary gain is achieved by avoiding the scaling altogether when the scale factor is 1.0.

Because the math is not exactly the same, some scaled pixel values may change by +- 1 RGBW "step".  In practice, this is unlikely to matter.

The gains are bigger on the Pico and other M0 microcontrollers than M4 microcontrollers with floating point math in the hardware.

Happily, flash size is also improved a bit on the Pico build I did, going from
> 542552 bytes used, 506024 bytes free in flash firmware space out of 1048576 bytes (1024.0kB).
    
to
> 542376 bytes used, 506200 bytes free in flash firmware space out of 1048576 bytes (1024.0kB).

### CODE REVIEW

File: shared-module/_pixelbuf/PixelBuf.c [hunk 1]
```diff
@@ -109,26 +111,31 @@ void common_hal__pixelbuf_pixelbuf_set_brightness(mp_obj_t self_in, mp_float_t b
...
+        if (self->pre_brightness_buffer == NULL) {
+            self->pre_brightness_buffer = m_malloc(pixel_len, false);
+            memcpy(self->pre_brightness_buffer, self->post_brightness_buffer, pixel_len);
+        }
+        for (size_t i = 0; i < pixel_len; i++) {
+            // Don't adjust per-pixel luminance bytes in dotstar mode
+            if (self->byteorder.is_dotstar && i % 4 == 0) {
+                continue;
+            }
+            self->post_brightness_buffer[i] = (self->pre_brightness_buffer[i] * self->scaled_brightness) >> 8;
```
Comment (tannewt) [inline shared-module/_pixelbuf/PixelBuf.c:133]: Division is clearer than a shift and should result in equivalent code.

```suggestion
            self->post_brightness_buffer[i] = (self->pre_brightness_buffer[i] * self->scaled_brightness) / 256;
```

File: shared-module/_pixelbuf/PixelBuf.c [hunk 2]
```diff
@@ -109,26 +111,31 @@ void common_hal__pixelbuf_pixelbuf_set_brightness(mp_obj_t self_in, mp_float_t b
     pixelbuf_pixelbuf_obj_t* self = native_pixelbuf(self_in);
     // Skip out if the brightness is already set. The default of self->brightness is 1.0. So, this
     // also prevents the pre_brightness_buffer allocation when brightness is set to 1.0 again.
-    mp_float_t change = brightness - self->brightness;
-    if (-0.001 < change && change < 0.001) {
+    self->brightness = brightness;
+    uint16_t new_scaled_brightness = (int)roundf(brightness * 256);
```
Comment (tannewt) [inline shared-module/_pixelbuf/PixelBuf.c:115]:
```suggestion
    // Use 256 steps of brightness so that we can do integer math below.
    uint16_t new_scaled_brightness = (int)roundf(brightness * 256);
```

File: shared-module/_pixelbuf/PixelBuf.c [hunk 3]
```diff
@@ -318,3 +331,23 @@ void common_hal__pixelbuf_pixelbuf_fill(mp_obj_t self_in, mp_obj_t fill_color) {
         common_hal__pixelbuf_pixelbuf_show(self_in);
     }
 }
+
+mp_int_t common_hal__pixelbuf_get_buffer(mp_obj_t self_in, mp_buffer_info_t *bufinfo, mp_uint_t flags) {
```
Comment (tannewt) [inline shared-module/_pixelbuf/PixelBuf.c:335]: Why is this added? I don't see it used anywhere.

File: shared-module/_pixelbuf/PixelBuf.c [hunk 4]
```diff
@@ -197,28 +204,34 @@ void _pixelbuf_set_pixel_color(pixelbuf_pixelbuf_obj_t* self, size_t index, uint
...
+        if (self->bytes_per_pixel == 4) {
+            if (!self->byteorder.is_dotstar) {
+                w = (w * self->scaled_brightness) >> 8;
+            }
+            scaled_buffer[rgbw_order->w] = w;
         }
-        post_brightness_buffer[rgbw_order->w] = w;
+        scaled_buffer[rgbw_order->r] = (r * self->scaled_brightness) >> 8;
+        scaled_buffer[rgbw_order->g] = (g * self->scaled_brightness) >> 8;
+        scaled_buffer[rgbw_order->b] = (b * self->scaled_brightness) >> 8;
```
Comment (tannewt) [inline shared-module/_pixelbuf/PixelBuf.c:233]: Divide here too.

**Review by tannewt — CHANGES_REQUESTED** [review]
Thanks for the speed up! Just a couple small things.

File: shared-module/_pixelbuf/PixelBuf.c (hunk 3 above)
Comment (jepler) [reply shared-module/_pixelbuf/PixelBuf.c]: oops -- making a pixelbuf memoryview'able is a separate thing..

**Review by tannewt — APPROVED** [review]
Looks great! Thank you!
